lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/StochasticGradientDescent.java

```diff
@@ -183,29 +214,39 @@ public String getOutputType() {
     /**
      * Trains the learning algorithm given an object as an example.
      *
-     * @param exampleFeatures The example's array of feature indices.
-     * @param exampleValues The example's array of feature values.
-     * @param exampleLabels The example's label(s).
-     * @param labelValues The labels' values.
+     * @param exampleFeatures  The example's array of feature indices.
+     * @param exampleValues    The example's array of feature values.
+     * @param exampleLabels    The example's label(s).
+     * @param labelValues      The labels' values.
      **/
-    public void learn(int[] exampleFeatures, double[] exampleValues, int[] exampleLabels,
-            double[] labelValues) {
-        assert exampleLabels.length == 1 : "Example must have a single label.";
+    public void learn(int[] exampleFeatures, double[] exampleValues,
+                      int[] exampleLabels, double[] labelValues) {
+        assert exampleLabels.length == 1
+                : "Example must have a single label.";
 
         double labelValue = labelValues[0];
-        double multiplier =
-                learningRate
-                        * (labelValue - weightVector.dot(exampleFeatures, exampleValues) - bias);
-        weightVector.scaledAdd(exampleFeatures, exampleValues, multiplier);
-        bias += multiplier;
+        double wtx = weightVector.dot(exampleFeatures, exampleValues) + bias;
+
+        if (isLMS) {
+            double multiplier = learningRate * (labelValue - wtx);
+            weightVector.scaledAdd(exampleFeatures, exampleValues, multiplier);
+            bias += multiplier;
+        }
+        else {
+            if (labelValue * wtx <= 1) {
+                double multiplier = learningRate * labelValue;
```

Comment:
For now it's ok, but as I suggested earlier the implementation of loss functions can be more general. You can define a trait/abstract class for loss functions, with functions to get the value and gradient, given an input and output label; and implement it for each instance of the loss function, e.g. LMS, hinge, etc.